pkg/deps/cargo.go

```diff
@@ -0,0 +1,154 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package deps
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+
+	"github.com/apache/skywalking-eyes/internal/logger"
+)
+
+type CargoMetadata struct {
+	Packages []CargoPackage `json:"packages"`
+}
+
+type CargoPackage struct {
+	Name         string `json:"name"`
+	Version      string `json:"version"`
+	License      string `json:"license"`
+	LicenseFile  string `json:"license_file"`
+	ManifestPath string `json:"manifest_path"`
+}
+
+type CargoTomlResolver struct {
+	Resolver
+}
+
+func (resolver *CargoTomlResolver) CanResolve(file string) bool {
+	base := filepath.Base(file)
+	logger.Log.Debugln("Base name:", base)
+	return base == "Cargo.toml"
+}
+
+// Resolve resolves licenses of all dependencies declared in the Cargo.toml file.
+func (resolver *CargoTomlResolver) Resolve(cargoTomlFile string, config *ConfigDeps, report *Report) error {
+	dir := filepath.Dir(cargoTomlFile)
+
+	download := exec.Command("cargo", "fetch")
+	logger.Log.Debugf("Run command: %v, please wait", download.String())
+	download.Stdout = os.Stdout
+	download.Stderr = os.Stderr
+	download.Dir = dir
+	if err := download.Run(); err != nil {
+		return err
+	}
+
+	cmd := exec.Command("cargo", "metadata", "--format-version=1", "--all-features")
+	cmd.Dir = dir
+	output, err := cmd.Output()
+	if err != nil {
+		return err
+	}
+
+	var metadata CargoMetadata
+	if err := json.Unmarshal(output, &metadata); err != nil {
+		return err
+	}
+
+	logger.Log.Debugln("Package size:", len(metadata.Packages))
+
+	return resolver.ResolvePackages(metadata.Packages, config, report)
+}
+
+// ResolvePackages resolves the licenses of the given packages.
+func (resolver *CargoTomlResolver) ResolvePackages(packages []CargoPackage, config *ConfigDeps, report *Report) error {
+	for i := range packages {
+		pkg := packages[i]
+
+		if config.IsExcluded(pkg.Name, pkg.Version) {
+			continue
+		}
+		if l, ok := config.GetUserConfiguredLicense(pkg.Name, pkg.Version); ok {
+			report.Resolve(&Result{
+				Dependency:    pkg.Name,
+				LicenseSpdxID: l,
+				Version:       pkg.Version,
+			})
+			continue
+		}
+		err := resolver.ResolvePackageLicense(config, &pkg, report)
+		if err != nil {
+			logger.Log.Warnf("Failed to resolve the license of <%s@%s>: %v\n", pkg.Name, pkg.Version, err)
+			report.Skip(&Result{
+				Dependency:    pkg.Name,
+				LicenseSpdxID: Unknown,
+				Version:       pkg.Version,
+			})
+		}
+	}
+	return nil
+}
+
+var cargoPossibleLicenseFileName = regexp.MustCompile(`(?i)^LICENSE|LICENCE(\.txt)?|LICENSE-.+|COPYING(\.txt)?$`)
+
+// ResolvePackageLicense resolve the package license.
+// The CargoPackage.LicenseFile is generally used for non-standard licenses and is ignored now.
+func (resolver *CargoTomlResolver) ResolvePackageLicense(config *ConfigDeps, pkg *CargoPackage, report *Report) error {
+	if pkg.License == "" {
+		return fmt.Errorf("license is empty")
+	}
```

Comment:
Should be this?

```suggestion
	if pkg.License != "" {
		return pkg.License
	}
```

Reply:
I wrote it with reference to `GoModResolver`, which returns an error, the calling place will warn and call `report.Skip`

Reply:
Hi @jmjoy , I mean if `pkg.License != ""` we can just use `pkg.License` (no need to read its license file content)

Reply:
@kezhenxu94  In fact, I don't need the --output parameter, but I saw that this parameter originally existed, so I implemented it.

Reply:
OK. Then we don't need this `if` statement. This will basically report a false negative case if the `pkg.license` field is empty but there is a license file in the crate

Reply:
@jmjoy I think you might want to apply this patch

```diff
diff --git a/pkg/deps/cargo.go b/pkg/deps/cargo.go
index d1057ab..e2613d2 100644
--- a/pkg/deps/cargo.go
+++ b/pkg/deps/cargo.go
@@ -19,13 +19,13 @@ package deps
 
 import (
 	"encoding/json"
-	"fmt"
 	"os"
 	"os/exec"
 	"path/filepath"
 	"regexp"
 
 	"github.com/apache/skywalking-eyes/internal/logger"
+	"github.com/apache/skywalking-eyes/pkg/license"
 )
 
 type CargoMetadata struct {
@@ -114,10 +114,6 @@ var cargoPossibleLicenseFileName = regexp.MustCompile(`(?i)^LICENSE|LICENCE(\.tx
 // ResolvePackageLicense resolve the package license.
 // The CargoPackage.LicenseFile is generally used for non-standard licenses and is ignored now.
 func (resolver *CargoTomlResolver) ResolvePackageLicense(config *ConfigDeps, pkg *CargoPackage, report *Report) error {
-	if pkg.License == "" {
-		return fmt.Errorf("license is empty")
-	}
-
 	dir := filepath.Dir(pkg.ManifestPath)
 	logger.Log.Debugf("Directory of %+v is %+v", pkg.Name, dir)
 	files, err := os.ReadDir(dir)
@@ -128,6 +124,8 @@ func (resolver *CargoTomlResolver) ResolvePackageLicense(config *ConfigDeps, pkg
 	var licenseFilePath string
 	var licenseContent []byte
 
+	licenseID := pkg.License
+
 	for _, info := range files {
 		if !cargoPossibleLicenseFileName.MatchString(info.Name()) {
 			continue
@@ -142,11 +140,17 @@ func (resolver *CargoTomlResolver) ResolvePackageLicense(config *ConfigDeps, pkg
 		break
 	}
 
+	if licenseID == "" { // If pkg.License is empty, identify the license ID from the license file content
+		if licenseID, err = license.Identify(string(licenseContent), config.Threshold); err != nil {
+			return err
+		}
+	}
+
 	report.Resolve(&Result{
 		Dependency:      pkg.Name,
 		LicenseFilePath: licenseFilePath,
 		LicenseContent:  string(licenseContent),
-		LicenseSpdxID:   pkg.License,
+		LicenseSpdxID:   licenseID,
 		Version:         pkg.Version,
 	})
 
```

Reply:
Thanks, it's better, just push to the patch to my branch? Because I open the `Allow edits and access to secrets by maintainers`.

Reply:
> Thanks, it's better, just push the patch to my branch? Because I open the `Allow edits and access to secrets by maintainers`.

Yes. Feel free to just `git apply` my patch and push in this PR. Or if you want me to edit I can do that too.

Reply:
@kezhenxu94 Done.